File: packages/dmn-marshaller/tests/jbang/dmnValidation.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import * as fs from "fs";
import * as path from "path";
import { getMarshaller } from "@kie-tools/dmn-marshaller";
import { fail } from "assert";
import { executeJBangScript } from "./jbangManager";

/**
 * This test suite validates the xml produced (parsed and built) by the marshaller relying on KIE DMN Validator
 * (https://github.com/apache/incubator-kie-drools/tree/main/kie-dmn/kie-dmn-validation).
 * A JBang script is used to actually call the KIE DMN Validator Java code.
 */

const dmnTestingModelsPath = require.resolve("@kie-tools/dmn-testing-models");

const dmnTestingModels = [
  "../valid_models/DMNv1_5/AllowedValuesChecksInsideCollection.dmn",
  "../valid_models/DMNv1_5/DateToDateTimeFunction.dmn",
  "../valid_models/DMNv1_5/ForLoopDatesEvaluate.dmn",
  "../valid_models/DMNv1_5/Imported_Model_Unamed.dmn",
  "../valid_models/DMNv1_5/ListReplaceEvaluate.dmn",
  "../valid_models/DMNv1_5/NegationOfDurationEvaluate.dmn",
  "../valid_models/DMNv1_5/TypeConstraintsChecks.dmn",
  "../valid_models/DMNv1_x/multiple/Financial.dmn",
  "../valid_models/DMNv1_x/multiple/Imported_Traffic_Violation.dmn",
  "../valid_models/DMNv1_x/multiple/stdlib.dmn",
  "../valid_models/DMNv1_x/allTypes.dmn",
  "../valid_models/DMNv1_x/dtevent.dmn",
  "../valid_models/DMNv1_x/habitability.dmn",
  "../valid_models/DMNv1_x/loan.dmn",
  "../valid_models/DMNv1_x/LoanEligibility.dmn",
  "../valid_models/DMNv1_x/OneOfEachType.dmn",
  "../valid_models/DMNv1_x/Prequalification.dmn",
  "../valid_models/DMNv1_x/testWithExtensionElements.dmn",
  "../valid_models/DMNv1_x/Traffic Violation Simple.dmn",
  "../valid_models/DMNv1_x/Traffic Violation.dmn",
];

const dmnTestingImportedModels = [
  {
    imported: "../valid_models/DMNv1_5/Imported_Model_Unamed.dmn",
    importer: "../valid_models/DMNv1_5/Importing_EmptyNamed_Model_With_Href_Namespace.dmn",
  },
  {
    imported: "../valid_models/DMNv1_5/Imported_Model_Unamed.dmn",
    importer: "../valid_models/DMNv1_5/Importing_EmptyNamed_Model_Without_Href_Namespace.dmn",
  },
  {
    imported: "../valid_models/DMNv1_5/Imported_Model_Unamed.dmn",
    importer: "../valid_models/DMNv1_5/Importing_Named_Model.dmn",
  },
  {
    imported: "../valid_models/DMNv1_5/Imported_Model_Unamed.dmn",
    importer: "../valid_models/DMNv1_5/Importing_OverridingEmptyNamed_Model.dmn",
  },
  {
    imported: "../valid_models/DMNv1_x/multiple/Imported_Traffic_Violation.dmn",
    importer: "../valid_models/DMNv1_x/multiple/Traffic Violation With Import.dmn",
  },
];
export const dmnValidationGeneratedFilesDirectory = path.join(
  __dirname,
  "../../dist-tests/dmnValidation-generated-files"
);
const scriptPath = path.join(__dirname, "./DmnValidation.java");

export function executeValidationTests() {
  for (const file of dmnTestingModels) {
    testFile(path.join(dmnTestingModelsPath, file));
  }
  for (const file of dmnTestingImportedModels) {
    file.imported = path.join(dmnTestingModelsPath, file.imported);
    file.importer = path.join(dmnTestingModelsPath, file.importer);
    testImportedFile(file);
  }
}

function testFile(normalizedFsPathRelativeToTheFile: string) {
  test(
    "DMN Validation: " +
      normalizedFsPathRelativeToTheFile.substring(normalizedFsPathRelativeToTheFile.lastIndexOf(path.sep) + 1),
    () => {
      const generatedXMLFilePath = parseXMLAndWriteInFile(normalizedFsPathRelativeToTheFile);

      try {
        executeJBangScript(scriptPath, "-d" + generatedXMLFilePath);
      } catch (error) {
        const fileName = normalizedFsPathRelativeToTheFile.substring(
          normalizedFsPathRelativeToTheFile.lastIndexOf(path.sep) + 1
        );
        fail("Validation of " + fileName + " failed! Please scroll up the logs to see the reason.");
      }
    }
  );
}

function testImportedFile(normalizedFsPathRelativeToTheFiles: { imported: string; importer: string }) {
  test(
    "DMN Validation: " +
      normalizedFsPathRelativeToTheFiles.importer.substring(
        normalizedFsPathRelativeToTheFiles.importer.lastIndexOf(path.sep) + 1
      ),
    () => {
      const importedGeneratedXMLFilePath = parseXMLAndWriteInFile(normalizedFsPathRelativeToTheFiles.imported);
      const importerGeneratedXMLFilePath = parseXMLAndWriteInFile(normalizedFsPathRelativeToTheFiles.importer);

      try {
        executeJBangScript(scriptPath, "-d" + importedGeneratedXMLFilePath, "-i" + importerGeneratedXMLFilePath);
      } catch (error) {
        const fileName = normalizedFsPathRelativeToTheFiles.importer.substring(
          normalizedFsPathRelativeToTheFiles.importer.lastIndexOf(path.sep) + 1
        );
        fail("Validation of " + fileName + " failed! Please scroll up the logs to see the reason.");
      }
    }
  );
}

function parseXMLAndWriteInFile(normalizedFsPathRelativeToTheFile: string): string {
  const originalXML = fs.readFileSync(normalizedFsPathRelativeToTheFile, "utf-8");
  const { parser, builder } = getMarshaller(originalXML, { upgradeTo: "latest" });
  const generatedXML = builder.build(parser.parse());
  const fileName = normalizedFsPathRelativeToTheFile.substring(
    normalizedFsPathRelativeToTheFile.lastIndexOf(path.sep) + 1
  );
  const generatedXMLFilePath = dmnValidationGeneratedFilesDirectory + path.sep + fileName;
  fs.writeFileSync(generatedXMLFilePath, generatedXML, "utf-8");
  return generatedXMLFilePath;
}

File: packages/dmn-marshaller-backend-compatibility-tester/src/DmnValidation.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

///usr/bin/env jbang "$0" "$@" ; exit $?
//SOURCES ./DmnMarshallerBackendCompatibilityTesterScript.java
/* Please do not declare new dependencies here, but in the above Parent class to preserve 
 *  the dependencies fetching mechanism */

package jbang;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

import org.kie.dmn.api.core.DMNMessage;
import org.kie.dmn.core.compiler.profiles.ExtendedDMNProfile;
import org.kie.dmn.validation.DMNValidator;
import org.kie.dmn.validation.DMNValidator.Validation;
import org.kie.dmn.validation.DMNValidatorFactory;

/**
 * JBang script that performs DMN files' XML (in string format) validation
 * relying on KIE DMN Validator
 * (https://github.com/apache/incubator-kie-drools/tree/main/kie-dmn/kie-dmn-validation).
 * The script can manage one or two (in case of imported model) DMN file paths.
 * The XSD SCHEMA, DMN COMPLIANCE and DMN COMPILATION are validated.
 */
@Command(name = "DmnSemanticComparison", mixinStandardHelpOptions = true, version = "DmnSemanticComparison 0.1", description = "It validates given DMN files")
class DmnValidation extends DmnMarshallerBackendCompatibilityTesterScript {

    private static final Logger LOGGER = LoggerFactory.getLogger(DmnValidation.class);

    @Option(names = {"-c", "--command"}, description = "Command to execute (no_imports) or (with_imports)", required = true)
    private String command;

    @Option(names = {"-d", "--dmnFilePath"}, description = "Path of DMN file to be validated", required = true)
    private String dmnFilePath;

    @Option(names = {"-i", "--importedDmnFilesPaths"} , description = "Paths of the DMN files imported by the DMN file to validate", required = false, split = ",")
    private String[] importedDmnFilesPath;

    public static void main(String... args) {
        int exitCode = new CommandLine(new DmnValidation()).execute(args);
        System.exit(exitCode);
    }

    @Override
    public Integer call() {
        try {
            switch (command) {
                case "no_imports":
                    return validateDMNModelsNoImports();
                case "with_imports": {
                    if (importedDmnFilesPath == null || importedDmnFilesPath.length == 0) {
                        throw new IllegalArgumentException("Imported DMN paths are missing");
                    }
                    return validateDMNModelsWithImports();
                }
                default:
                    LOGGER.error("Unknown command {}", command);
                    return 1;
            }
        } catch (Exception e) {
            LOGGER.error("Failed to execute {}", command, e);
            return 100;
        }
    }

    private int validateDMNModelsNoImports() {
        File dmnFile = new File(dmnFilePath);
        DMNValidator dmnValidator = DMNValidatorFactory.newValidator(List.of(new ExtendedDMNProfile()));
        final List<DMNMessage> messages = dmnValidator.validateUsing(Validation.VALIDATE_SCHEMA,
                Validation.VALIDATE_MODEL,
                Validation.VALIDATE_COMPILATION)
                .theseModels(dmnFile);
        return assessDMNMessageResults(messages, dmnFile.getName());
    }

    private int validateDMNModelsWithImports() {
        File dmnFile = new File(dmnFilePath);
        List<File> models = new ArrayList<>();
        models.add(dmnFile);
        models.addAll(Stream.of(importedDmnFilesPath)
                .map(File::new)
                .collect(Collectors.toList()));
        DMNValidator dmnValidator = DMNValidatorFactory.newValidator(List.of(new ExtendedDMNProfile()));
        List<DMNMessage> messages = dmnValidator.validateUsing(
                Validation.VALIDATE_SCHEMA,
                Validation.VALIDATE_MODEL,
                Validation.VALIDATE_COMPILATION)
                .theseModels(models.toArray(File[]::new));
        return assessDMNMessageResults(messages, dmnFile.getName());
    }

    private int assessDMNMessageResults(List<DMNMessage> dmnMessageResults, String dmnFileName) {
        if (dmnMessageResults.size() == 0) {
            LOGGER.info("RESULT: Following files have been successfully validated!");
            return 0;
        } else {
            LOGGER.error("ERROR: DMN Validation failed for the DMN file " + dmnFileName);
            LOGGER.error("Validation Errors:");
            List<String> messages = dmnMessageResults.stream().map(DMNMessage::getText).collect(Collectors.toList());
            messages.forEach(LOGGER::error);
            System.err.println("ERROR: DMN Validation failed!");
            System.err.println("DMN File Name: " + dmnFileName);
            messages.forEach(System.err::println);
            return 1;
        }
    }
}